KAFKA-13835: Fix two bugs related to dynamic broker configs in KRaft

cmccabe opened the pull request:
The first bug is that we are calling reloadUpdatedFilesWithoutConfigChange when a topic
configuration is changed, but not when a broker configuration is changed. This is backwards.
This function must be called only for BROKER configs, and never for TOPIC configs. (Also, this
function is called only for specific broker configs, not for cluster configs.)

The second bug is that there were several configurations such as `max.connections` which are
related to broker listeners, but which do not involve creating or removing new listeners. We should
support these configurations in KRaft. Only adding or removing listeners is unsupported, not
changes to existing listeners. This PR fixes the configuration change validation to support this.
(This also removes a very puzzling error mesage stating that we are adding or removing listeners,
when we are not).

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -906,17 +906,22 @@ class DynamicClientQuotaCallback(server: KafkaBroker) extends Reconfigurable {
-  def validateReconfiguration(newConfig: KafkaConfig): Unit = {
-    val oldConfig = server.config
-    if (!oldConfig.requiresZookeeper) {
+  def checkIsNotKRaft(): Unit = {
```
nit: could this be private?

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -906,17 +906,22 @@ class DynamicClientQuotaCallback(server: KafkaBroker) extends Reconfigurable {
-    if (!oldConfig.requiresZookeeper) {
+  def checkIsNotKRaft(): Unit = {
+    if (!broker.config.requiresZookeeper) {
       throw new ConfigException("Dynamic reconfiguration of listeners is not yet supported when using a Raft-based metadata quorum")
```
It seems a little strange to throw such a specific exception for a generically named method like `checkIsNotKRaft`. How about `verifyDynamicListenersSupported()` or something like that?

cmccabe replied:
changed to `verifyAdditionOrRemovalIsSupported` which is more specific. I also updated the exception text to be more specific.

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -950,16 +963,17 @@ class DynamicListenerConfig(server: KafkaBroker) extends BrokerReconfigurable wi
-    server match {
-      case kafkaServer: KafkaServer => kafkaServer.kafkaController.updateBrokerInfo(kafkaServer.createBrokerInfo)
-      case _ =>
+    LoginManager.closeAll()
```
Do we still want to do this even if we reject the configuration change?

cmccabe replied:
Let me look into this more...

cmccabe replied:
I fixed this so that we only call this if an endpoint was added or removed, which is how it was previously.

hachikuji commented on core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala:
```diff
@@ -248,6 +248,24 @@ class BrokerMetadataListener(
+
+    override def run(): Unit = {
+      _publisher = Some(publisher)
+      log.info(s"Set publisher to ${publisher}")
```
Why info? Seems more like a debug level message.

cmccabe replied:
This is only used in unit tests. I think it's good to have it in the log4j since earlier events may have been published differently by the earlier publisher.

mdedetrich replied:
If this is only used in tests does it make sense to alter `alterPublisher` to give it a second parameter accepting a value that implements `EventQueue.FailureLoggingEvent` and then define the `AlterPublisherEvent` class in `/core/src/test` rather than `core/src/main`.

That way its more clear the class is only being used for tests.

cmccabe replied:
@mdedetrich : Thanks for the comment. I would like to put this code outside of `BrokerMetadataListener`, but that would mean making `_publisher` public (or at least non-private) which is something I really don't want to do. People have a tendency to use whatever variables are available to them in junit tests, even if there are concurrency problems. We have a lot of examples of that. This function is more like a setter. It's simple enough to just put in the code.

hachikuji commented on core/src/main/scala/kafka/server/metadata/BrokerMetadataPublisher.scala:
```diff
@@ -187,21 +189,26 @@ class BrokerMetadataPublisher(conf: KafkaConfig,
+              if (resource.name().isEmpty) {
+                // Apply changes to "cluster configs" (also known as default BROKER configs).
+                // These are stored in KRaft with an empty name field.
+                info(s"Updating cluster configuration : " +
```
nit: either drop the `s` prefix before the string or move `toLoggableProps` to an `${}` substitution

hachikuji commented on core/src/main/scala/kafka/server/metadata/BrokerMetadataPublisher.scala:
```diff
@@ -187,21 +189,26 @@ class BrokerMetadataPublisher(conf: KafkaConfig,
+                  toLoggableProps(resource, props).mkString(","))
+                dynamicConfigHandlers(ConfigType.Broker).
+                  processConfigChanges(ConfigEntityName.Default, props)
+              } else if (resource.name().equals(brokerId.toString)) {
```
nit: in scala, we can use `==`

cmccabe replied:
Hmm, do we need to do it that way? `equals` seems clearer since it means the same thing in Java.

If you really feel like it's non-idiomatic then I can change...

mdedetrich replied:
I echo this, in Scala pretty much **no one** uses `equals`. Scala deliberately overrides `==` so that it always uses `equals` and that is the normal way of doing non reference object comparison.

cmccabe replied:
OK, I will change it to == here to make it more idiomatic Scala code.

hachikuji commented on core/src/test/scala/unit/kafka/server/metadata/BrokerMetadataPublisherTest.scala:
```diff
@@ -142,4 +157,80 @@ class BrokerMetadataPublisherTest {
+  {
+    val numTimesReloadCalled = new AtomicInteger(0)
+
+    override def reloadUpdatedFilesWithoutConfigChange(props: Properties): Unit = {
```
It feels like we could probably do this with a Mockito spy.

cmccabe replied:
done

hachikuji reviewed (commented):
Nice finds. Left a few small comments.

cmccabe committed:
KAFKA-13835: Fix two bugs related to dynamic broker configs in KRaft #12063

The first bug is that we are calling reloadUpdatedFilesWithoutConfigChange when a topic
configuration is changed, but not when a broker configuration is changed. This is backwards.  This
function must be called only for BROKER configs, and never for TOPIC configs. (Also, this function
is called only for specific broker configs, not for cluster configs.)

The second bug is that there were several configurations such as max.connections which are related
to broker listeners, but which do not involve creating or removing new listeners. We should support
these configurations in KRaft. Only adding or removing advertised listeners is unsupported, not
changes to existing listeners. This PR fixes the configuration change validation to support this.

cmccabe commented:
I pushed some changes. Sorry that this is a rebase. I was testing something out and it got annoying to do it as a merge.

Still looking into a test failure in `DynamicBrokerReconfigurationTest.testAdvertisedListenerUpdate` which is very odd considering how little the ZK path is changed here. Timing dependent?

cmccabe commented:
I had to re-fix this a bit. It turns out that my PR had a bug where if an advertised listener was modified (but not removed or added) we would not update the registration in ZK. This should be fixed now.

This also further clarifies the KRaft check : even removing or adding listeners is OK, as long as advertised listeners are not touched.

cmccabe commented:
The only test failure I could find was the "heap space exhausted" thing which has nothing to do with this PR.

cmccabe committed:
equals to ==

cmccabe committed:
Merge branch 'trunk' into KAFKA-13835

hachikuji commented on core/src/main/scala/kafka/server/DynamicBrokerConfig.scala:
```diff
@@ -948,18 +980,18 @@ class DynamicListenerConfig(server: KafkaBroker) extends BrokerReconfigurable wi
-    server match {
-      case kafkaServer: KafkaServer => kafkaServer.kafkaController.updateBrokerInfo(kafkaServer.createBrokerInfo)
-      case _ =>
+    if (listenersRemoved.nonEmpty || listenersAdded.nonEmpty) {
```
Could this check be moved inside the `listenerRegistrationsAltered` block below? I'm trying to understand why they handled separately.

cmccabe replied:
`listenerRegistrationsAltered` is about *registered listeners*, which may be different from *listeners*.

One example is if you have a registered listener set to `PLAINTEXT://myhostname:9092`. Changing that to `PLAINTEXT://myhostname2:9092` is a registered listener change which is not possible in KRaft currently.

But, you could have the *listener* change from `PLAINTEXT://0.0.0.0:9092` (listen on all interfaces) to `PLAINTEXT://myhostname:9092` (listen only on the IP associated with myhostname). *That* would be possible in kraft.

And of course, we only want to alter the znode in ZK mode if the registered listeners changed.

hachikuji reviewed (approved):
Thanks, LGTM

cmccabe commented:
Jenkins test failures were `ConnectorTopicsIntegrationTest` and `ConnectionQuotasTest`, which are not related and which pass locally. Committing. Thanks, all.
